File: DoraDB/src/storage/disk_manager.cpp
#include "storage/disk_manager.h"
#include "common/config.h"

#include <fcntl.h>      // open()
#include <unistd.h>     // read(), write(), close(), lseek()
#include <sys/stat.h>   // file permissions
#include <cstring>      // memset
#include <stdexcept>
#include <iostream>
#include <filesystem>

DiskManager::DiskManager(const std::string& file_path) : file_path_(file_path) {
    // Make sure the parent directory exists
    std::filesystem::path p(file_path);
    if (p.has_parent_path()) {
        std::filesystem::create_directories(p.parent_path());
    }

    // Open or create the file (read-write, create if missing)
    fd_ = open(file_path.c_str(), O_RDWR | O_CREAT, 0644);
    if (fd_ < 0) {
        throw std::runtime_error("DiskManager: cannot open file " + file_path);
    }

    // Figure out how many pages already exist by checking file size
    off_t file_size = lseek(fd_, 0, SEEK_END);
    num_pages_ = file_size / PAGE_SIZE;
}

DiskManager::~DiskManager() {
    if (fd_ >= 0) {
        close(fd_);
        fd_ = -1;
    }
}

void DiskManager::ReadPage(int page_id, char* data) {
    if (page_id < 0 || page_id >= num_pages_) {
        throw std::runtime_error("DiskManager::ReadPage: invalid page_id " +
                                 std::to_string(page_id));
    }

    off_t offset = static_cast<off_t>(page_id) * PAGE_SIZE;
    lseek(fd_, offset, SEEK_SET);

    ssize_t bytes_read = read(fd_, data, PAGE_SIZE);
    if (bytes_read != PAGE_SIZE) {
        throw std::runtime_error("DiskManager::ReadPage: short read on page " +
                                 std::to_string(page_id));
    }
}

void DiskManager::WritePage(int page_id, const char* data) {
    if (page_id < 0) {
        throw std::runtime_error("DiskManager::WritePage: invalid page_id");
    }

    off_t offset = static_cast<off_t>(page_id) * PAGE_SIZE;
    lseek(fd_, offset, SEEK_SET);

    ssize_t bytes_written = write(fd_, data, PAGE_SIZE);
    if (bytes_written != PAGE_SIZE) {
        throw std::runtime_error("DiskManager::WritePage: short write on page " +
                                 std::to_string(page_id));
    }
}

int DiskManager::AllocatePage() {
    int new_page_id = num_pages_;
    num_pages_++;

    // Extend the file by writing one empty page
    char empty[PAGE_SIZE];
    memset(empty, 0, PAGE_SIZE);
    WritePage(new_page_id, empty);

    return new_page_id;
}

File: MiniDB_Projects/Team_OrionDB/src/storage/disk_manager.py
import os

class DiskManager:
    PAGE_SIZE = 4096

    def __init__(self, db_filepath):
        self.db_filepath = db_filepath
        # Open in r+b mode if exists, otherwise w+b to create it
        if os.path.exists(db_filepath):
            self.file = open(db_filepath, "r+b")
        else:
            self.file = open(db_filepath, "w+b")
        self.num_pages = self._get_num_pages_on_disk()

    def _get_num_pages_on_disk(self):
        self.file.seek(0, os.SEEK_END)
        file_size = self.file.tell()
        return file_size // self.PAGE_SIZE

    def write_page(self, page_id, page_data):
        assert len(page_data) == self.PAGE_SIZE, f"Page data size must be {self.PAGE_SIZE}"
        self.file.seek(page_id * self.PAGE_SIZE)
        self.file.write(page_data)
        self.file.flush()

    def read_page(self, page_id):
        self.file.seek(page_id * self.PAGE_SIZE)
        data = self.file.read(self.PAGE_SIZE)
        if len(data) < self.PAGE_SIZE:
            # Return an empty page if read is incomplete or out of bounds
            return bytearray(self.PAGE_SIZE)
        return bytearray(data)

    def allocate_page(self):
        new_page_id = self.num_pages
        # Write an empty page to extend the file
        empty_data = bytearray(self.PAGE_SIZE)
        self.write_page(new_page_id, empty_data)
        self.num_pages += 1
        return new_page_id

    def close(self):
        if self.file and not self.file.closed:
            self.file.close()
